### ISSUE TITLE
8.1905.0 parameters missing

### ISSUE THREAD

**Issue by DagOveH** [rsyslog/rsyslog#3692]
I just upgraded to 8.1905.0 from 8.1901.0, but now I see new errors;
Jun 05 15:19:43 <node> rsyslogd[23930]: error during config processing: parameter 'action.reportsuspension' not known -- typo in config file? [v8.1905.0 try https://www.rsyslog.com/e/2207 ]
Jun 05 15:19:43 <node> rsyslogd[23930]: error during config processing: parameter 'action.resumeRetryCount' not known -- typo in config file? [v8.1905.0 try https://www.rsyslog.com/e/2207 ]
Jun 05 15:19:43 <node> rsyslogd[23930]:  [origin software="rsyslogd" swVersion="8.1905.0" x-pid="23930" x-info="https://www.rsyslog.com"] start

These parameters is used in a omkafka action block and in an main_queue() block.

A search in the Changelog did not give information of these changes ...

**Comment by rgerhards** [comment]
Config?

**Comment by DagOveH** [comment]
Uhh, sorry, config pasted below.

Further investigation shows these 2 error messages are from the main_queue() block.
If I comment out these in that block
  action.resumeRetryCount="-1"
  action.reportsuspension="on"

... the error messages disappears.
(I have therefore removed (omkafka) from the title)

Platform: 
Red Hat Enterprise Linux Server release 7.6 (Maipo)

Config; (
#### GLOBAL DIRECTIVES ####
#
# Where to place auxiliary files
global(workDirectory="/data/disk01/rsyslog_etl/dns_v2")
#
# Use default timestamp format
#$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$ActionFileDefaultTemplate RSYSLOG_SyslogProtocol23Format
#
# Set to use IP v4 only for the imudp/imtcp modules
global(net.ipprotocol="ipv4-only")
#
# Configures the maximum message size allowed for all inputs.
# For mesages over 4k its recomended to use TCP and not UDP
global(maxMessageSize="64k")
#global(maxDataSize="64k")
#
#### END GLOBAL DIRECTIVES ####

#### INCLUDE TEMPLATES ####
include(file="/etc/rsyslog_etl/common.templates" mode="abort-if-missing")
include(file="/etc/rsyslog_etl/dns_v2.templates" mode="abort-if-missing")
#### INCLUDE TEMPLATES ####

#### SOC ETL Config begin ####
# Modules to use
# dohe 30.01.2019: Performance tuning to ProcessOnPoller to "off"
# dohe 30.01.2019: Performance tuning to increase Threads from default 2 to 4 for high-traffic environments
# dohe 22.05.2019: Decreased Threads to 2 agains, 4 was not needed. Max for Threads is number of CPU's minus 1
module(load="impstats" resetCounters="on" format="json-elasticsearch" interval="60" ruleset="rsyslog_metric")
module(load="imptcp" Threads="2" ProcessOnPoller="off")
module(load="mmnormalize")
module(load="mmutf8fix")
module(load="omkafka")
module(load="omelasticsearch")

# dohe 31.01.2019: Performance tuning to set RateLimit.Burst to "0" and disable flowControl (Sender will not be affected)
# dohe 31.01.2019: Performance tuning to set flowControl to "off" to help sender when they cant buffer/queue correct (DNS server)
input(type="imptcp" RateLimit.Burst="0" port="60055" flowControl="off" ruleset="dns_v2")

ruleset(name="dns_v2") {
  action(type="mmnormalize" ruleBase="/etc/rsyslog_etl/dns_v2.rb" useRawMsg="on")
  # USE OK or FAIL for $parsesuccess
  if $parsesuccess == "OK" then {
    action(type="mmutf8fix")
    action(type="omkafka"
      name="dns_v2_to_kafka_v2_action"
      broker=[
        "removed:9093",
        "removed:9093",
        "removed:9093"
      ]
      confParam=[
        "compression.codec=gzip",
        "api.version.request=true",
        "socket.timeout.ms=9000",
        "socket.keepalive.enable=true",
        "security.protocol=ssl",
        "ssl.ca.location=removed",
        "ssl.certificate.location=removed",
        "ssl.key.location=removed",
        "ssl.key.password=removed"
      ]
      dynatopic="off"
      failedMsgFile="/data/disk01/rsyslog_etl/dns_v2/kafka2_failed.msg"
      keepFailedMessages="on"
      resubmitOnFailure="on"
      Partitions.Auto="on"
      closeTimeout="10000"
      template="json-kafka2"
      topic="dns_v2"
      queue.spoolDirectory="/data/disk01/rsyslog_etl/dns_v2"
      queue.filename="omkafka_dns_v2_to_kafka_v2_action.queue"
      queue.type="LinkedList"
      queue.saveonshutdown="on"
      queue.MaxFileSize="1G"
      queue.maxDiskSpace="100G"
      action.ResumeRetryCount="-1"
      action.reportSuspension="on"
      queue.size="500000"
      queue.workerThreads="5"
      queue.dequeueBatchSize="100000"
      queue.minDequeueBatchSize="10000"
      queue.minDequeueBatchSize.timeout="10000"
    )
    #action(type="omfile" ziplevel="0" template="json-kafka2" file="/data/disk01/rsyslog_etl/dns_v2/out.json")
  }
  else if $parsesuccess == "FAIL" then {
    action(type="omfile" ziplevel="0" template="MyForwardTemplate" file="/data/disk01/rsyslog_etl/dns_v2/failed_parsing_events.txt")
    action(type="omfile" ziplevel="0" template="RSYSLOG_DebugFormat" file="/data/disk01/rsyslog_etl/dns_v2/failed_parsing_events_debug.txt")
  }
}

main_queue(
  queue.type="LinkedList"
  queue.dequeueBatchSize="10000"   # numbers of messages to be parsed from queue
#  queue.highWatermark="450000"    # max no. of events to hold in memory
#  queue.lowWatermark="250000"     # use memory queue again, when it's back to this level
#  dohe: Not needed, as default is 90% and 70% of queue.size
  queue.spoolDirectory="/data/disk01/rsyslog_etl/dns_v2"
  queue.fileName="rsyslog_main_queue"
  queue.maxDiskSpace="100g"        # it will stop at this much disk space
  queue.saveOnShutdown="on"      # save memory queue contents to disk when rsyslog is exiting
  queue.MaxFileSize="1G"
  queue.MaxDiskSpace="100G"
  action.resumeRetryCount="-1"
  action.reportsuspension="on"
  queue.size="500000"
)

ruleset(name="rsyslog_metric"){
  set $!etl = "dns_v2";   # Local variables are local to the current message
  action(
    name="parse_rsyslog_stats"
    type="mmnormalize"
    ruleBase="/etc/rsyslog_etl/rsyslog_metric_json.rb" # contains just: rule=:%data:json%
  )
  action(
    name="push_rsyslog_metric"
    type="omelasticsearch"
    action.resumeInterval="60"
    bulkmode="on"
    errorFile="/data/disk01/rsyslog_etl/dns_v2/omelasticsearch_error.msg"
    server=[
      "removed",
      "removed",
      "removed",
      "removed",
      "removed",
      "removed",
      "removed",
      "removed",
      "removed",
      "removed"
    ]
    serverport="9200"
    template="rsyslog_metric_template"
    searchIndex="rsyslog_etl-index"
    dynSearchIndex="on"
    queue.spoolDirectory="/data/disk01/rsyslog_etl/dns_v2"
    queue.filename="omelasticsearch.queue"
    queue.type="LinkedList"
    queue.saveonshutdown="on"
    queue.maxFileSize="100m"
    action.resumeRetryCount="-1"
  )
}

#### ETL Config end ####

**Comment by davidelang** [comment]
On Wed, 5 Jun 2019, Dag Ove Hegreberg wrote:

> Further investigation shows these 2 error messages are from the main_queue() block.
> If I comment out these in that block
>  action.resumeRetryCount="-1"
>  action.reportsuspension="on"

action.* parameters make no sense when applied to the main_queue(), which is why 
they were generating errors.

additional comments inline below

> ... the error messages disappears.
> (I have therefore removed (omkafka) from the title)
>
> Platform:
> Red Hat Enterprise Linux Server release 7.6 (Maipo)
>
> Config; (
> #### GLOBAL DIRECTIVES ####
> #
> # Where to place auxiliary files
> global(workDirectory="/data/disk01/rsyslog_etl/dns_v2")
> #
> # Use default timestamp format
> #$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
> $ActionFileDefaultTemplate RSYSLOG_SyslogProtocol23Format
> #
> # Set to use IP v4 only for the imudp/imtcp modules
> global(net.ipprotocol="ipv4-only")
> #
> # Configures the maximum message size allowed for all inputs.
> # For mesages over 4k its recomended to use TCP and not UDP
> global(maxMessageSize="64k")

you really should have all the global items in one block, I don't know if 
multiple blocks with different items listed will combine them of have a 'last 
one defined wins' approach.

global (net.ipprotocol="ipv4-only" maxMessageSize="64k")

would be the correct way to do this.

> #global(maxDataSize="64k")
> #
> #### END GLOBAL DIRECTIVES ####
>
> #### INCLUDE TEMPLATES ####
> include(file="/etc/rsyslog_etl/common.templates" mode="abort-if-missing")
> include(file="/etc/rsyslog_etl/dns_v2.templates" mode="abort-if-missing")
> #### INCLUDE TEMPLATES ####
>
> #### SOC ETL Config begin ####
> # Modules to use
> # dohe 30.01.2019: Performance tuning to ProcessOnPoller to "off"
> # dohe 30.01.2019: Performance tuning to increase Threads from default 2 to 4 for high-traffic environments
> # dohe 22.05.2019: Decreased Threads to 2 agains, 4 was not needed. Max for Threads is number of CPU's minus 1
> module(load="impstats" resetCounters="on" format="json-elasticsearch" interval="60" ruleset="rsyslog_metric")
> module(load="imptcp" Threads="2" ProcessOnPoller="off")
> module(load="mmnormalize")
> module(load="mmutf8fix")
> module(load="omkafka")
> module(load="omelasticsearch")
>
> # dohe 31.01.2019: Performance tuning to set RateLimit.Burst to "0" and disable flowControl (Sender will not be affected)
> # dohe 31.01.2019: Performance tuning to set flowControl to "off" to help sender when they cant buffer/queue correct (DNS server)
> input(type="imptcp" RateLimit.Burst="0" port="60055" flowControl="off" ruleset="dns_v2")
>
> ruleset(name="dns_v2") {
>  action(type="mmnormalize" ruleBase="/etc/rsyslog_etl/dns_v2.rb" useRawMsg="on")
>  # USE OK or FAIL for $parsesuccess
>  if $parsesuccess == "OK" then {
>    action(type="mmutf8fix")
>    action(type="omkafka"
>      name="dns_v2_to_kafka_v2_action"
>      broker=[
>        "removed:9093",
>        "removed:9093",
>        "removed:9093"
>      ]
>      confParam=[
>        "compression.codec=gzip",
>        "api.version.request=true",
>        "socket.timeout.ms=9000",
>        "socket.keepalive.enable=true",
>        "security.protocol=ssl",
>        "ssl.ca.location=removed",
>        "ssl.certificate.location=removed",
>        "ssl.key.location=removed",
>        "ssl.key.password=removed"
>      ]
>      dynatopic="off"
>      failedMsgFile="/data/disk01/rsyslog_etl/dns_v2/kafka2_failed.msg"
>      keepFailedMessages="on"
>      resubmitOnFailure="on"
>      Partitions.Auto="on"
>      closeTimeout="10000"
>      template="json-kafka2"
>      topic="dns_v2"
>      queue.spoolDirectory="/data/disk01/rsyslog_etl/dns_v2"
>      queue.filename="omkafka_dns_v2_to_kafka_v2_action.queue"
>      queue.type="LinkedList"
>      queue.saveonshutdown="on"
>      queue.MaxFileSize="1G"
>      queue.maxDiskSpace="100G"
>      action.ResumeRetryCount="-1"
>      action.reportSuspension="on"
>      queue.size="500000"
>      queue.workerThreads="5"
>      queue.dequeueBatchSize="100000"
>      queue.minDequeueBatchSize="10000"
>      queue.minDequeueBatchSize.timeout="10000"
>    )
>    #action(type="omfile" ziplevel="0" template="json-kafka2" file="/data/disk01/rsyslog_etl/dns_v2/out.json")
>  }
>  else if $parsesuccess == "FAIL" then {
>    action(type="omfile" ziplevel="0" template="MyForwardTemplate" file="/data/disk01/rsyslog_etl/dns_v2/failed_parsing_events.txt")
>    action(type="omfile" ziplevel="0" template="RSYSLOG_DebugFormat" file="/data/disk01/rsyslog_etl/dns_v2/failed_parsing_events_debug.txt")
>  }
> }
>
> main_queue(
>  queue.type="LinkedList"
>  queue.dequeueBatchSize="10000"   # numbers of messages to be parsed from queue
> #  queue.highWatermark="450000"    # max no. of events to hold in memory
> #  queue.lowWatermark="250000"     # use memory queue again, when it's back to this level
> #  dohe: Not needed, as default is 90% and 70% of queue.size

you cannot put comments in the middle of a parameter list.

this is logically the same thing as having something like foo( bar=1 baz=2 ) and 
thinking you can do foo( #bar=1 baz=2)

>  queue.spoolDirectory="/data/disk01/rsyslog_etl/dns_v2"
>  queue.fileName="rsyslog_main_queue"
>  queue.maxDiskSpace="100g"        # it will stop at this much disk space
>  queue.saveOnShutdown="on"      # save memory queue contents to disk when rsyslog is exiting
>  queue.MaxFileSize="1G"
>  queue.MaxDiskSpace="100G"
>  action.resumeRetryCount="-1"
>  action.reportsuspension="on"
>  queue.size="500000"
> )
>
> ruleset(name="rsyslog_metric"){
>  set $!etl = "dns_v2";   # Local variables are local to the current message
>  action(
>    name="parse_rsyslog_stats"
>    type="mmnormalize"
>    ruleBase="/etc/rsyslog_etl/rsyslog_metric_json.rb" # contains just: rule=:%data:json%
>  )
>  action(
>    name="push_rsyslog_metric"
>    type="omelasticsearch"
>    action.resumeInterval="60"
>    bulkmode="on"
>    errorFile="/data/disk01/rsyslog_etl/dns_v2/omelasticsearch_error.msg"
>    server=[
>      "removed",
>      "removed",
>      "removed",
>      "removed",
>      "removed",
>      "removed",
>      "removed",
>      "removed",
>      "removed",
>      "removed"
>    ]
>    serverport="9200"
>    template="rsyslog_metric_template"
>    searchIndex="rsyslog_etl-index"
>    dynSearchIndex="on"
>    queue.spoolDirectory="/data/disk01/rsyslog_etl/dns_v2"
>    queue.filename="omelasticsearch.queue"
>    queue.type="LinkedList"
>    queue.saveonshutdown="on"
>    queue.maxFileSize="100m"
>    action.resumeRetryCount="-1"
>  )
> }
>
> #### ETL Config end ####
>
>
>

**Comment by rgerhards** [comment]
As David said: The `main_queue()` object does not have these parameters, consequently they are rejected. I don't know why previous versions seem to have ignored them silently. If so, that was a bug. It was probably fixed as a side-effect of some other change.

**Comment by rgerhards** [comment]
> main_queue( queue.type="LinkedList" queue.dequeueBatchSize="10000" # numbers of messages to be parsed from queue # queue.highWatermark="450000" # max no. of events to hold in memory # queue.lowWatermark="250000" # use memory queue again, when it's back to this level # dohe: Not needed, as default is 90% and 70% of queue.size
> you cannot put comments in the middle of a parameter list.

actually, in this case you can. These are `line comments`. The config processor understand this (commonly used) format. Everything in a line after `#` is a comment and ignored. No matter if inside an object definition or not. The same also applies to `/*` and `*/` if properly nested. 

Tech info: comments are stripped very early in config processing. It is done via the lexer file while tokenizing.

**Comment by davidelang** [comment]
On Thu, 6 Jun 2019, Rainer Gerhards wrote:

>> main_queue( queue.type="LinkedList" queue.dequeueBatchSize="10000" # numbers of messages to be parsed from queue # queue.highWatermark="450000" # max no. of events to hold in memory # queue.lowWatermark="250000" # use memory queue again, when it's back to this level # dohe: Not needed, as default is 90% and 70% of queue.size
>> you cannot put comments in the middle of a parameter list.
>
> actually, in this case you can. These are `line comments`. The config processor understand this (commonly used) format. Everything in a line after `#` is a comment and ignored. No matter if inside an object definition or not. The same also applies to `/*` and `*/` if properly nested.
>
> Tech info: comments are stripped very early in config processing. It is done via the lexer file while tokenizing.

good to know, thanks.

David Lang

**Comment by DagOveH** [comment]
I'm sorry for raising this without checking more myself.
I'm relatively new to rsyslog, being testing/using rsyslog to Kafka for some months now.
The config I have ended up with is the result of hours searching the web for information and examples.
I do appreciate the feedback from both @davidelang and @rgerhards and will use those to improve the setup.
One positive thing with the new version, I did see in the 8.1905.0 that saved DA queues files after rsyslog was stopped, is now picked up again when rsyslog is started.
Keep up the good work 😊
